### PR TITLE
Add TRADEOFFS.md

### PR DESCRIPTION
**Pull request by jquense**
The content here is mostly copied/paraphrased from @gaearon from a [twitter thread](https://twitter.com/dan_abramov/status/909180238856323072) the other day. I saw that Babel [had an issue open](https://github.com/babel/babel/issues/4959) adding a Tradeoffs.md I thought it would be great thing to add to the React repo as well. Maybe we want to move it to the docs directly but I thought i’d get something started, spur some conversation, and make this more of a community endeavor. 

I intentionally left the “pros” sparse for the time being  since I don’t think we’ll all have much trouble enumerating them :)

### CODE REVIEW

**Comment by bjacobel** [comment]
This may work better as a contribution to the [existing Design Principles document](https://facebook.github.io/react/contributing/design-principles.html).

**Comment by gaearon** [comment]
Yea, I think it might be a good idea to either make Design Principles more balanced with downsides, or to move that content here.

**Comment by jquense** [comment]
That could work, my original thought was that Design Principles is fairly wordy, vs a more straightforward pro/con like this. I think it's additionally helpful to have a frank list of tradeoffs for folks to look that, but i'm not opposed to rolling it into a more prose-y format :)

File: TRADEOFFS.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+
+## Potential reasons to use React
+
+- React has a HUGE communitity! If you have a problem or need odds are you'll be able to find someone who shares your needs and may already have a solution!
+- Create React App! Is a great way to start a React project without needing to futz around with tooling.
+
+## Potential reasons to not use React
+
+- Most React apps typically require a build step, this isn't technically required, but you'll find that the vast majority of React apps have one, and all learning material assumes one.
```
Comment (Kovensky) [inline TRADEOFFS.md]: However, the overwhelmingly majority of serious web projects will have _a_ build step anyway, if only to minify code 🤔

File: TRADEOFFS.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Tradeoffs
+
+
+## Potential reasons to use React
+
+- React has a HUGE communitity! If you have a problem or need odds are you'll be able to find someone who shares your needs and may already have a solution!
+- Create React App! Is a great way to start a React project without needing to futz around with tooling.
```
Comment (MichaelDeBoey) [inline TRADEOFFS.md]: Maybe add a link to the repo?

File: TRADEOFFS.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+## Potential reasons to not use React
+
+- Most React apps typically require a build step, this isn't technically required, but you'll find that the vast majority of React apps have one, and all learning material assumes one.
+- It is JavaScript-centric. If you aren't interested in learning JavaScript well, React may be harder to work with than template based libraries.
+- Larger runtime than other options. For projectes sensitive to byte size, such as embedded widgets, React may be too big
```
Comment (probablyup) [inline TRADEOFFS.md]: Typo: projectes

File: TRADEOFFS.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+- React's performance is worse than template-based solutions for cases where many things update extremely quickly at the same time, such as a stock trading app.
```
Comment (viankakrisna) [inline TRADEOFFS.md]: is this is always the case? can we elaborate more about this?

File: TRADEOFFS.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+- React trades memory pressure for expressiveness, meaning React apps typically do more short lived allocations under heavy load.
+- React perfers predictability over conciseness, which results in more time spent explicitly "wiring" code together.
```
Comment (viankakrisna) [inline TRADEOFFS.md]: perfers -> prefers

File: TRADEOFFS.md (hunk 4 above)
Comment (cztomsik) [reply TRADEOFFS.md]: I think this is because you can compile template directly into DOM calls, so there is no overhead of diffing (of course the performance is always tricky but in general, compiled templates are faster - see marko for example)

Comment (viankakrisna) [reply TRADEOFFS.md]: I see, so in other words, react needs to traverse the component tree and compare the difference every time it renders, and a compiled template can be optimized ahead of time to just modify the DOM directly? For example:

```
<div>
  <div>
    {props.children}
  </div>
</div>
```

if we want to update the props.children,

In React this would be a minimum of 2 createElement calls at runtime and then apply the difference to the DOM (two steps that happen for every render)

In an AoT template, this can be just one step (to put it simply), because the compiler knows that the only thing that needs to be updated is the `props.children`. Is that right?

Comment (cztomsik) [reply TRADEOFFS.md]: Yes :)

Comment (Kovensky) [reply TRADEOFFS.md]: This also affects the initial render time, because you first must build the `createElement` tree, then diff it against the DOM (which should be missing everything), then commit it.

This is mitigated by doing server side rendering, of course.

File: TRADEOFFS.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+- React has a HUGE communitity! It can be hard to find up-to-date, high quality, information in such a big space.
+- React's development is mostly driven by the needs of Facebook, if your apps are very different from what facebook is building it may not meet your needs well.
```
Comment (strugee) [inline TRADEOFFS.md]: Drive-by nit: capitalize "facebook"

**Comment by gaearon** [comment]
This is cool but from a glance it sounds like we're recommending people not to use React. :-)

I think something more nuanced could work, but it's probably better to do on the website than in this repository. For example as a "Tradeoffs" page that goes into more detail, or by amending the "Design Principles".
